### PR TITLE
3.1 handle symlink init

### PR DESCRIPTION
**Pull request by c-nixon**
This updates the Tailer to handle initialization of symlinks where their target files that are outside the tracked folders

### CODE REVIEW

File: common/state/src/lib.rs [hunk 1]
```diff
@@ -98,6 +98,12 @@ impl AgentState {
 #[derive(Hash, Clone, PartialEq, Eq)]
 pub struct FileName(bytes::Bytes);
 
+impl FileName {
+    pub fn bytes(&self) -> &[u8] {
```
Comment (jorgebay) [inline common/state/src/lib.rs:102]: I think this is not used.

**Review by jorgebay — COMMENTED** [review]
It's looking good, I'll give it a try on k8s w/ lookback enabled.

File: common/fs/src/tail.rs [hunk 2]
```diff
@@ -106,49 +106,124 @@ impl Tailer {
...
+                                            debug!("File is in initial dirs");
+                                            data.borrow_mut().tail(vec![path]).await
+                                        } else {
+                                            None
+                                        }
+                                    }
+                                    Entry::Symlink { name, link, .. } => {
+                                        let sym_path = path.clone();
+                                        let sym_name = name;
+                                        info!("Initialise event for symlink {:?}, target {:?}", name, link);
```
Comment (jorgebay) [inline common/fs/src/tail.rs:160]: We should unify `Initialise` vs `initialize` :) 

Messages in logs show and mixed casing, for example:

```
[2021-03-26T10:41:09Z INFO  fs::tail] initialized "/var/lib/docker/containers/f527c4bac8f3d84606a21d08016227f989a7fe61f19107c9324ff1ea4426dba6/f527c4bac8f3d84606a21d08016227f989a7fe61f19107c9324ff1ea4426dba6-json.log" with offset 34274
[2021-03-26T10:41:09Z INFO  fs::tail] Initialise event for symlink "7.log", target "/var/lib/docker/containers/f527c4bac8f3d84606a21d08016227f989a7fe61f19107c9324ff1ea4426dba6/f527c4bac8f3d84606a21d08016227f989a7fe61f19107c9324ff1ea4426dba6-json.log"
```

File: common/fs/src/cache/mod.rs [hunk 3]
```diff
@@ -150,7 +150,14 @@ impl FileSystem {
                 } else {
                     for event in events {
                         match event {
-                            Event::New(entry) => fs.initial_events.push(Event::Initialize(entry)),
+                            Event::New(entry_key) => {
+                                if let Some(entry) = entries.get(entry_key) {
+                                    let path = fs.resolve_direct_path(entry, &entries);
+                                    if fs.is_initial_dir_target(&path) {
+                                        fs.initial_events.push(Event::Initialize(entry_key))
+                                    }
```
Comment (jorgebay) [inline common/fs/src/cache/mod.rs:158]: Makes sense to only emit events for the initial target files 👍

File: bin/tests/cli.rs [hunk 4]
```diff
@@ -441,6 +442,7 @@ fn test_append_after_symlinks_delete() {
 #[cfg_attr(not(feature = "integration_tests"), ignore)]
 #[cfg_attr(not(target_os = "linux"), ignore)]
 fn test_directory_symlinks_delete() {
+    let _ = env_logger::Builder::from_default_env().try_init();
```
Comment (jorgebay) [inline bin/tests/cli.rs:445]: NIT: We can review all this logging output.

File: bin/tests/cli.rs [hunk 5]
```diff
@@ -777,6 +781,351 @@ async fn test_tags() {
...
+    let (server_result, _) = tokio::join!(server, async {
+        let settings = AgentSettings::with_mock_ingester(&log_dir.to_str().unwrap(), &addr);
+        let mut agent_handle = common::spawn_agent(settings);
+        let stderr_reader = std::io::BufReader::new(agent_handle.stderr.take().unwrap());
+        // Consume output
+        std::thread::spawn(move || {
+            stderr_reader.lines().for_each(|line| {
+                eprintln!("---- line: {}", line.unwrap());
+            });
+        });
```
Comment (jorgebay) [inline bin/tests/cli.rs:960]: Same: Let's replace it with a call to `consume_output()`.

File: bin/tests/cli.rs [hunk 6]
```diff
@@ -777,6 +781,351 @@ async fn test_tags() {
...
+        tokio::time::delay_for(tokio::time::Duration::from_millis(2000)).await;
+        for i in 10..20 {
+            writeln!(file, "SAMPLE {}", i).unwrap();
+        }
+        file.sync_all().unwrap();
+        common::force_client_to_flush(&log_dir).await;
+        // Wait for the data to be received by the mock ingester
+        tokio::time::delay_for(tokio::time::Duration::from_millis(2000)).await;
+        let map = received.lock().await;
+        eprintln!("--TEST MAP KEYS: {:?}", map.keys());
```
Comment (jorgebay) [inline bin/tests/cli.rs:1026]: NIT: :)

**Review by jorgebay — COMMENTED** [review]
Looks great, let's remove all the unnecessary test output and ship it 👍

**Comment by jorgebay** [comment]
BTW, tested it on k8s with a counter pod and works perfectly with and w/o stateful lookback.

File: common/state/src/lib.rs (hunk 1 above)
Comment (c-nixon) [reply common/state/src/lib.rs:102]: I was using it to print out the contents. I've replaced it with derive(Debug) on the type

**Review by jorgebay — APPROVED** [review]
Nice!

**Pull request merged**